### DISCUSSION TITLE
How to lazy load neotest

### DISCUSSION THREAD

**Discussion by rajneesh2k10** [nvim-neotest/neotest#455] [Q&A] [answered]
Here is my current lazy config for neotest -

```
{
  "nvim-neotest/neotest",
  dependencies = {
    "nvim-neotest/nvim-nio",
    "nvim-lua/plenary.nvim",
    "antoinemadec/FixCursorHold.nvim",
    "nvim-treesitter/nvim-treesitter",
  },
  config = function()
    require("neotest").setup({
      adapters = {
        require("rustaceanvim.neotest"),
      },
    })
  end,
}
```

After adding this config, I've been seeing OCCASIONAL error on opening ANY files (non-rust files too) -

```
Failed to run `config` for neotest

...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:80: Cannot find package "codelldb".
```

My guess is that, mason (from lsp-config) & neotest (or its dependencies) get into a race of loading themselves. Depending on who wins, I some times get errors and sometimes don't.

Two questions -

1. How can I prevent this race by lazy loading neotest as I don't need it at the time of starting vim?

2. Neotest requires adapters for different languages, when opening a Rust file, I don't want to load the Go language adapter (also the Go LSP) and vice-versa. #428 

TIA

**Comment by fredrikaverpil** [comment, marked as answer]
There are a few different ways to do that. I just add `keys` and assign my desired keymaps. This will resulting in not loading the plugin until I hit one of those keymaps. 

```diff
{
  "nvim-neotest/neotest",
  dependencies = {
    "nvim-neotest/nvim-nio",
    "nvim-lua/plenary.nvim",
    "antoinemadec/FixCursorHold.nvim",
    "nvim-treesitter/nvim-treesitter",
  },
  config = function()
    require("neotest").setup({
      adapters = {
        require("rustaceanvim.neotest"),
      },
    })
  end,
+ keys = {...},
}
```

EDIT: I've defined my keys in a separate function: https://github.com/fredrikaverpil/dotfiles/blob/fc670a75c7bd5a0349052fba3fd0183bb86a0fe1/nvim-fredrik/lua/config/keymaps.lua#L444-L535

Note that in this case you need to use anonymous functions (like I did) so that you won't load the plugin on sourcing of this code.

**Reply by rajneesh2k10** [reply]
Thanks for the response. This indeed helped! However, there is a little more catch to it. You need to define the keys in a way that it doesn't force loading of the neotest - 

```
	keys = {
		-- Run nearest tests
		vim.keymap.set("n", "<leader>R", function()
			require("neotest").run.run()
		end, { desc = "Run nearest tests" }),
		-- Run tests in file
		vim.keymap.set("n", "<leader>F", function()
			require("neotest").run.run(vim.fn.expand("%"))
		end, { desc = "Run tests in file" }),
	}
```

Instead of setting the keymap by directly requiring the "neotest" module, it has to be put behind a function (as shown in the config above).
Thanks to this Reddit post - https://www.reddit.com/r/neovim/comments/123ed7x/best_way_to_lazy_load_plugin_on_keys_that_include/

@fredrikaverpil - I've accepted your response as an answer. Could you please update your answer with the key definition?

**Reply by fredrikaverpil** [reply]
Sure, edited. I linked to my own config which does this. Sorry for not explaining this fully.
